llvm/test/TableGen/aie-variable-instr-itinerary.td

```diff
@@ -0,0 +1,132 @@
+
+class MOVInstruction : TestInstruction {
+  int Inst = 1;
+  let OutOperandList = (outs );
+  let InOperandList = (ins GPR:$val, i32imm:$adr);
+}
+
+let Itinerary = II_MOV  in
+  def MOV_instr : MOVInstruction<>;
+
+let Itinerary = II_MOV, ItineraryRegPairs = [ItinRegClassPair<II_MOV_GPR,GPR>]  in
+  def MOV_instr_GPR : MOVInstruction<>;
+
+let Itinerary = II_MOV, ItineraryRegPairs = [ItinRegClassPair<II_MOV_PTR,PTR>]  in
+  def MOV_instr_PTR : MOVInstruction<>;
+
+let Itinerary = II_MOV, ItineraryRegPairs = [ItinRegClassPair<II_MOV_GPR,GPR, 1>, ItinRegClassPair<II_MOV_PTR,PTR, 1>]  in
+  def MOV_instr_GPR_PTR : MOVInstruction<>;
+
+
+let Itinerary = II_MOV, ItineraryRegPairs = [ItinRegClassPair<II_MOV_GPR,GPR>]  in {
+  def MOV_instr_GPR_1 : MOVInstruction<>;
+  def MOV_instr_GPR_2 : MOVInstruction<>;  
+}
+
+let Itinerary = II_MOV, ItineraryRegPairs = [ItinRegClassPair<II_MOV_GPR,GPR, 1>, ItinRegClassPair<II_MOV_PTR,PTR, 1>]  in {
+  def MOV_instr_GPR_PTR_1 : MOVInstruction<>;
+  def MOV_instr_GPR_PTR_2 : MOVInstruction<>;
+}
+
+// CHECK:      static std::optional<unsigned> getOperandIdx(const unsigned Opcode) {
+// CHECK-NEXT:   switch (Opcode) {
+// CHECK-NEXT:   default:
+// CHECK-NEXT:     return {};
+// CHECK-NEXT:   case TestAIE::MOV_instr_GPR:
+// CHECK-NEXT:   case TestAIE::MOV_instr_GPR_1:
+// CHECK-NEXT:   case TestAIE::MOV_instr_GPR_2:
+// CHECK-NEXT:   case TestAIE::MOV_instr_PTR:
+// CHECK-NEXT:     return 0;
+// CHECK-NEXT:   case TestAIE::MOV_instr_GPR_PTR:
+// CHECK-NEXT:   case TestAIE::MOV_instr_GPR_PTR_1:
+// CHECK-NEXT:   case TestAIE::MOV_instr_GPR_PTR_2:
+// CHECK-NEXT:     return 1;
+// CHECK-NEXT:   }
+// CHECK-NEXT: }
+
+// CHECK:      unsigned TestAIEInstrInfo::getSchedClass(
+// CHECK-NEXT:     const MCInstrDesc &Desc,
+// CHECK-NEXT:     iterator_range<const MachineOperand *> Operands) const {
+// CHECK-NEXT:   std::optional<unsigned> OperandIdx = getOperandIdx(Desc.getOpcode());
+// CHECK-NEXT:   if (!OperandIdx.has_value())
+// CHECK-NEXT:     return Desc.getSchedClass();
+// CHECK-EMPTY: 
+// CHECK-NEXT:   const MachineOperand &MO = *std::next(Operands.begin(), OperandIdx.value());
+// CHECK-NEXT:   const Register Reg = MO.getReg();
+// CHECK-NEXT:   assert(MO.isReg() && "Operand must be of Reg type");
+// CHECK-EMPTY:  
+// CHECK-NEXT:   const MachineInstr *MI = MO.getParent();
+// CHECK-NEXT:   const MachineRegisterInfo &MRI =
+// CHECK-NEXT:       MI->getParent()->getParent()->getRegInfo();
+// CHECK-NEXT:   const TargetRegisterClass *RC =
+// CHECK-NEXT:       Reg.isVirtual() ? MRI.getRegClass(Reg) : nullptr;
+// CHECK-EMPTY: 
+// CHECK-NEXT:   switch (Desc.getOpcode()) {
+// CHECK-NEXT:   default:
+// CHECK-NEXT:     return Desc.getSchedClass();
+// CHECK-NEXT:   case TestAIE::MOV_instr_GPR:
+// CHECK-NEXT:   case TestAIE::MOV_instr_GPR_1:
+// CHECK-NEXT:   case TestAIE::MOV_instr_GPR_2:
+// CHECK-NEXT:   {
+// CHECK-NEXT:     if (AIEBaseInstrInfo::regClassMatches(TestAIE::GPRRegClass, RC, Reg))
```

Comment:
We could also just write the operand index here for each opcode. E.g. here we would write `checkRCForOperand(TestAIE::GPRRegClass, 0)`, and for `MOV_instr_GPR_PTR_2` we would write
`checkRCForOperand(TestAIE::GPRRegClass, 1)`.

Note I would define `checkRCForOperand()` as a simple lambda within this function, e.g.
```
auto checkRCForOperand= [&Operands, &MRI] (const TargetRegisterClass &ExpectedRC, unsigned OpIdx) {
  Register Reg = std::next(Operands.begin(), OpIdx)->getReg();
  const TargetRegisterClass *RC = Reg.isVirtual() ? MRI.getRegClass(Reg) : nullptr;
  return AIEBaseInstrInfo::regClassMatches(TestAIE::GPRRegClass, RC, Reg);
};
```

TLDR: I would be happy to have a single switch statement instead of two.